File: ios/Bonfire/Core/Channel/ChannelsDatabase.swift
import Foundation
import RxSwift

protocol ChannelsDatabase {
    func observePublicChannelIds() -> Observable<[String]>
    func observePrivateChannelIdsFor(user: User) -> Observable<[String]>
    func readChannelFor(channelName: String) -> Observable<Channel>
    func writeChannel(newChannel: Channel) -> Observable<Channel>
    func writeChannelToPublicChannelIndex(newChannel: Channel) -> Observable<Channel>
    func addOwnerToPrivateChannel(user: User, channel: Channel) -> Observable<Channel>
    func removeOwnerFromPrivateChannel(user: User, channel: Channel) -> Observable<Channel>
    func addChannelToUserPrivateChannelIndex(user: User, channel: Channel) -> Observable<Channel>
    func removeChannelFromUserPrivateChannelIndex(user: User, channel: Channel) -> Observable<Channel>
    func observeOwnerIdsFor(channel: Channel) -> Observable<[String]>
}

File: android/core/src/main/java/com/novoda/bonfire/channel/database/ChannelsDatabase.java
package com.novoda.bonfire.channel.database;

import com.novoda.bonfire.channel.data.model.Channel;
import com.novoda.bonfire.user.data.model.User;

import java.util.List;

import io.reactivex.Observable;

public interface ChannelsDatabase {

    Observable<List<String>> observePublicChannelIds();

    Observable<List<String>> observePrivateChannelIdsFor(User user);

    Observable<Channel> readChannelFor(String channelName);

    Observable<Channel> writeChannel(Channel newChannel);

    Observable<Channel> writeChannelToPublicChannelIndex(Channel newChannel);

    Observable<Channel> addOwnerToPrivateChannel(User user, Channel channel);

    Observable<Channel> removeOwnerFromPrivateChannel(User user, Channel channel);

    Observable<Channel> addChannelToUserPrivateChannelIndex(User user, Channel channel);

    Observable<Channel> removeChannelFromUserPrivateChannelIndex(User user, Channel channel);

    Observable<List<String>> observeOwnerIdsFor(Channel channel);

}
